### PR TITLE
add data augmentation and dc2 psf/deconv

### PR DESCRIPTION
**Pull request by XinyueLi1012**
1. add data augmentation 
2. add psf/deconv for dc2
will resolve https://github.com/prob-ml/bliss/issues/904 and https://github.com/prob-ml/bliss/issues/898

### CODE REVIEW

File: bliss/conf/base_config.yaml [hunk 1]
```diff
@@ -84,6 +84,9 @@ encoder:
         concat_psf_params: false
         log_transform: false
         rolling_z_score: true
+    data_augmentation:
+        do_data_augmentation: true
+        epoch_start: 1
```
Comment (jeff-regier) [inline bliss/conf/base_config.yaml:89]: Do we need the epoch_start option?

File: bliss/data_augmentation.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
+import random
+
+import numpy as np
+import torchvision as TF
```
Comment (jeff-regier) [inline bliss/data_augmentation.py:4]: Why TF?

File: bliss/data_augmentation.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    num_transform_list = [1, 2, 3]
+    num_transform = random.choices(num_transform_list, weights=[0.7, 0.2, 0.1], k=1)[0]
+    transform_list = ["vflip", "hflip", "rotate90", "rotate180", "rotate270", "shift"]
+
+    aug_method = np.random.choice(
+        transform_list, num_transform, p=(0.1, 0.1, 0.1, 0.1, 0.1, 0.5), replace=False
+    )
+
+    aug_image, aug_tile, aug_deconv = origin_full, image, deconv_image
+    for i in aug_method:
```
Comment (jeff-regier) [inline bliss/data_augmentation.py:22]: `i` usually denotes an integer. Would you use a more descriptive variable name here?

File: bliss/data_augmentation.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        if i == "vflip":
+            aug_image, aug_tile, aug_deconv = aug_vflip(aug_image, aug_tile, aug_deconv)
+        if i == "hflip":
+            aug_image, aug_tile, aug_deconv = aug_hflip(aug_image, aug_tile, aug_deconv)
+        if i == "rotate90":
+            aug_image, aug_tile, aug_deconv = aug_rotate90(aug_image, aug_tile, aug_deconv)
```
Comment (jeff-regier) [inline bliss/data_augmentation.py:28]: Can these lines be exercised by latest?

File: bliss/catalog.py [hunk 5]
```diff
@@ -488,6 +488,10 @@ def to_tile_params(
         for ii in range(self.batch_size):
             n_sources = int(self.n_sources[ii].item())
             for idx, coords in enumerate(tile_coords[ii][:n_sources]):
+                if coords[0] >= tile_n_sources.shape[1] or coords[1] >= tile_n_sources.shape[2]:
+                    continue
+                # ignore sources outside of the image (usually caused by data augmentation - shift)
```
Comment (jeff-regier) [inline bliss/catalog.py:493]: It seems preferable to crop the sources after the shift rather than doing it here

File: bliss/encoder.py [hunk 6]
```diff
@@ -330,6 +335,23 @@ def _get_loss(self, pred: Dict[str, Distribution], true_tile_cat: TileCatalog):
         return loss_with_components
 
     def _generic_step(self, batch, logging_name, log_metrics=False, plot_images=False):
+        do_data_augmentation = self.data_augmentation.get("do_data_augmentation")
+        data_augmentation_start = self.current_epoch >= self.data_augmentation.get("epoch_start")
+
+        if data_augmentation_start and (logging_name == "train") and do_data_augmentation:
+            deconv_image = None
+            if self.input_transform_params.get("use_deconv_channel"):
```
Comment (jeff-regier) [inline bliss/encoder.py:343]: Wouldn't we always want to rotate, shift, etc the deconvolution if one exists and we're rotating the main image?

**Review by jeff-regier — CHANGES_REQUESTED** [review]

File: bliss/conf/base_config.yaml (hunk 1 above)
Comment (jeff-regier) [reply bliss/conf/base_config.yaml:89]: i.e., why not always do data augmentation beginning with epoch 0 if augmentation is enabled?

File: bliss/conf/base_config.yaml [hunk 7]
```diff
@@ -227,7 +230,9 @@ surveys:
     dc2:
         _target_: bliss.surveys.dc2.DC2
         data_dir: /nfs/turbo/lsa-regier/lsstdesc-public/dc2/run2.2i-dr6-v4/coadd-t3828-t3829/deepCoadd-results/
-        cat_path: ${paths.dc2}/merged_catalog/merged_catalog_622.pkl
-        batch_size: 128
+        cat_path: ${paths.dc2}/merged_catalog/merged_catalog_psf_100.pkl
```
Comment (jeff-regier) [inline bliss/conf/base_config.yaml:233]: Is the code for creating these pkl files somewhere in our repo, e.g. the scripts directory?

File: bliss/data_augmentation.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+def augment_data(tile_catalog, image, deconv_image=None):
+    origin_tile = TileCatalog(4, tile_catalog)
+    origin_full = origin_tile.to_full_params()
+
+    num_transform_list = [1, 2, 3]
+    num_transform = random.choices(num_transform_list, weights=[0.7, 0.2, 0.1], k=1)[0]
+    transform_list = ["vflip", "hflip", "rotate90", "rotate180", "rotate270", "shift"]
+
+    aug_method = np.random.choice(
+        transform_list, num_transform, p=(0.1, 0.1, 0.1, 0.1, 0.1, 0.5), replace=False
```
Comment (jeff-regier) [inline bliss/data_augmentation.py:18]: We'd each of the eight rotation-flip pairs to be equally likely? Does this achieve that? Is there a simpler way to write it?

Comment (jeff-regier) [reply bliss/data_augmentation.py:18]: Don't we always want to call shift? With probability 1/16 it will shift by (0,0), effectively leaving the image unchanged, right?

File: bliss/conf/base_config.yaml (hunk 1 above)
Comment (jeff-regier) [reply bliss/conf/base_config.yaml:89]: It'd be best to remove `epoch_start` and to assume `epoch_start = 0`. It's easier on users to have few options to set, all else equal. Also, if we're using data augmentation, I think we'd want it from the beginning. Otherwise, we may overfit on some examples (the non-transformed ones) at first while underfitting on others.

File: bliss/encoder.py (hunk 6 above)
Comment (jeff-regier) [reply bliss/encoder.py:343]: Can the data augmentation be applied automatically to all the channels, i.e., whichever ones are present, without explicitly referencing any one channel? For example, the background channel should be rotated too if the image is rotated. It'd be nice if that were implemented without us having a special case for the background channel. Same from the deconv channel -- it doesn't seem like a special case should be necessary for it.

File: bliss/conf/base_config.yaml (hunk 7 above)
Comment (jeff-regier) [reply bliss/conf/base_config.yaml:233]: If possible, please include it in a case study as part of this PR.

File: bliss/data_augmentation.py (hunk 8 above)
Comment (jeff-regier) [reply bliss/data_augmentation.py:18]: I don't think the code as is gives us uniform coverage of rotation-flip pairs, which is what we want. It also seems a bit hard to understand.

This may be simpler way to do it: First, select one of four rotations (including 0 degrees -- a "no op") and apply. Next, select whether to flip the image vertically or not. (No need to every flip horizontally--that's the same as a vertical flip and a 180 degree rotation.) Finally, apply a random shift.

File: bliss/data_augmentation.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    plocs = origin_full["plocs"].clone()
+    origin_full["plocs"][:, :, 1] = image_size - plocs[:, :, 0] - 1
+    origin_full["plocs"][:, :, 0] = plocs[:, :, 1]
+    aug_tile = origin_full.to_tile_params(4, 4).get_brightest_source_per_tile().to_dict()
+    return aug_image, aug_tile, aug_deconv
+
+
+def aug_shift(origin_full, image, deconv_image):
+    shift_x = random.randint(0, 3)
+    shift_y = random.randint(0, 3)
```
Comment (jeff-regier) [inline bliss/data_augmentation.py:102]: Would it work to select a shift from `random.randint(-1, 2)` instead? I doubt it will make a big difference, but perhaps shifting by -1 rather than 3 seems less likely to truncate sources.

File: bliss/encoder.py [hunk 10]
```diff
@@ -330,6 +335,23 @@ def _get_loss(self, pred: Dict[str, Distribution], true_tile_cat: TileCatalog):
         return loss_with_components
 
     def _generic_step(self, batch, logging_name, log_metrics=False, plot_images=False):
+        do_data_augmentation = self.data_augmentation.get("do_data_augmentation")
```
Comment (jeff-regier) [inline bliss/encoder.py:338]: We only want to do data augmentation during training, never during validation or testing. Therefore, could we move the data augmentation call to `training_step(...)` rather than here?

File: bliss/encoder.py (hunk 6 above)
Comment (jeff-regier) [reply bliss/encoder.py:343]: In other words, how about concatenating all the images, background, deconv images, etc along the channel axis beforehand? The operations like `vflip` will operate on all of them automatically with just one call.

File: bliss/data_augmentation.py (hunk 2 above)
Comment (jeff-regier) [reply bliss/data_augmentation.py:4]: That is, `TF` seems like a potentially misleading name for `torchvision`.

How about instead doing `from torchvision.transforms import functional`?

File: bliss/catalog.py (hunk 5 above)
Comment (jeff-regier) [reply bliss/catalog.py:493]: OK sounds good

File: bliss/data_augmentation.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    return aug_image, origin_full
+
+
+def aug_rotate90(origin_full, image):
+    num_channel = image.size(2)
+    aug_images = []
+    for i in range(num_channel):
+        rotated_slice = TF.rotate(image[:, :, i, :, :], 90)
+        aug_images.append(rotated_slice)
+    aug_image = torch.stack(aug_images, dim=2)
```
Comment (jeff-regier) [inline bliss/data_augmentation.py:42]: The first 6+ lines of `aug_rotate90`, `aug_rotate180`, and `aug_rotate270` are duplicated code. Could you move them to a new helper function that takes an angle as an argument?

File: bliss/data_augmentation.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    plocs = origin_full["plocs"].clone()
+    origin_full["plocs"][:, :, 1] = plocs[:, :, 0]
+    origin_full["plocs"][:, :, 0] = image_size - plocs[:, :, 1] - 1
+    return aug_image, origin_full
+
+
+def aug_rotate180(origin_full, image):
+    num_channel = image.size(2)
+    aug_images = []
+    for i in range(num_channel):
```
Comment (jeff-regier) [inline bliss/data_augmentation.py:54]: Is it necessary to loop over the channels? Can you just call `TF.rotate` once for all the channel? (It may be necessary to combine dimensions 1 and 2 first using `rearrange`.)

File: bliss/data_augmentation.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+def augment_data(tile_catalog, image):
+    origin_tile = TileCatalog(4, tile_catalog)
+    origin_full = origin_tile.to_full_params()
+    aug_full, aug_image = origin_full, image
+
+    rotate_list = [None, aug_rotate90, aug_rotate180, aug_rotate270]
+    flip_list = [None, aug_vflip]
+    rotate_choice = random.choice(rotate_list)
+    flip_choice = random.choice(flip_list)
```
Comment (jeff-regier) [inline bliss/data_augmentation.py:17]: Looks good!

File: bliss/data_augmentation.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+def aug_shift(origin_full, image):
+    shift_x = random.randint(-1, 2)
+    shift_y = random.randint(-1, 2)
+    image_size = image.size(3)
+    image_lim = [2 - shift_x, 2 - shift_x + image_size, 2 - shift_y, 2 - shift_y + image_size]
+
+    num_channel = image.size(2)
+    aug_images = []
+    for i in range(num_channel):
```
Comment (jeff-regier) [inline bliss/data_augmentation.py:89]: Same here, it'd be nice not to loop over the channels. We should be able to pad all the channels at once.

File: bliss/encoder.py [hunk 15]
```diff
@@ -385,7 +409,8 @@ def _generic_step(self, batch, logging_name, log_metrics=False, plot_images=Fals
 
     def training_step(self, batch, batch_idx, optimizer_idx=0):
         """Training step (pytorch lightning)."""
-        return self._generic_step(batch, "train")
+        do_data_augmentation = self.do_data_augmentation
```
Comment (jeff-regier) [inline bliss/encoder.py:412]: Can you do the data augmentation here in `training_step` rather than in `_generic_step`? (We don't want to do data augmentation unless we're training.)

File: bliss/encoder.py (hunk 6 above)
Comment (jeff-regier) [reply bliss/encoder.py:343]: Sound good

File: bliss/data_augmentation.py (hunk 8 above)
Comment (jeff-regier) [reply bliss/data_augmentation.py]: Nice job with these changes

**Review by jeff-regier — APPROVED** [review]
Nice
